Write design patterns

cilim opened the pull request:
Task: [297](https://app.productive.io/1-infinum/tasks/1043925)

#### Aim

It would be nice to have a chapter that roughly describes what design patterns are and what are the most used ones in the Ruby world and that we use.

From a perspective of a developer I think that we'll use this chapter when we're faced with situations where we're planning a feature and aren't sure what the right approach is.
That's why we'll need a structured way to make this chapter helpful and easy to find what you need.

I skimmed through the **Design patterns in Ruby** book by _Russ Olsen_, read a couple of articles and went through all the design patterns on [Refactoring Guru](https://refactoring.guru/design-patterns) and the Ruby community is pretty synched in regards to the design pattern naming conventions (with the exception of different Factory patterns - we usually use the Simple Factory)

#### Solution

I summarized most of the Design patterns that can be found on [Refactoring Guru](https://refactoring.guru/design-patterns). It's a very nice site and all the examples are explained with great detail. There's also Ruby examples so I don't think we need to dig in much further than that, and our Handbook chapter will just be a quick reference to it, that helps developers find what they need (with the **Problem** sections).

I deployed the chapter to our [staging environment](https://beta.infinum.com/handbook/books/rails/development-practices/design-patterns).
Check it out and please help me decide on:
* should I continue giving a small example of the design pattern as I did in the first 2 chapters?
* did I miss any design patterns?
* should I include more specific design patterns that belong to families of the main patterns described in this chapter? (eg, **Null Object** pattern is just an improvement to error handling when dealing with the **Factory Method**)

d4be4st commented:
omg i did not know about refactoring.guru website. looks awesome

My two cents:
* Would not include code examples, because then you would have to keep them updated. Would rather have different links with multiple examples.
* Some of those patterns i have never used like Prototype
* Different names than we usually use. What would be a Form object?
* Missing are some of the ones i most use: Service (or Command as guru calls it), State
* Rails specific patterns are missing like Repository, Form ...

cilim commented:
Yeah, refactoring.guru is definitely something I'm planning on researching in depth :)

> Would not include code examples, because then you would have to keep them updated. Would rather have different links with multiple examples

that's a valid point, but since it's a design pattern, I thought there's little chance of us having to maintain it, cause I don't see it changing 🤔 

> Different names than we usually use

Maybe we should get on board with the community? :D 

> What would be a Form object?

I was thinking about that and I would place it in the Creational Design Pattern category, but if you try and search anything more on form objects outside the context of Rails, I can't find any results.
Since it makes sense to keep it because we use it on a daily basis, I'll find some article that's not from refactoring.guru and reference it.

> Missing are some of the ones i most use: Service (or Command as guru calls it), State

I'll add the `Command` and the Rails specific ones then (State [is here](https://github.com/infinum/rails-handbook/pull/117/files#diff-f5cea16a5f762555e1d1efbf4c751f96R192)).

cilim commented:
@d4be4st all done:

- added table of contents
- added missing design patterns
- restructured and added references

Changes can be found [here](https://beta.infinum.com/handbook/books/rails/development-practices/design-patterns).

d4be4st commented:
Can you try with details html tag to make it more readable?

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/details

cilim commented:
@d4be4st what exactly do you want hidden?

d4be4st commented:
The content. was thinking of just showing the titles which are expandable.

cilim commented:
I don't think that's a good way to go, cause then you have to click on every design pattern to see what it solves and does it suit you and your problem :)

But that's just my opinion :) Maybe @infinum/backend-team can jump in too :) (hopefully everyone has Infinum VPN :D )

cilim commented:
@melcha @d4be4st I had to extend the CSS a bit for the collapsing details to work.

The result can be seen in the chapter: https://beta.infinum.com/handbook/books/rails/development-practices/design-patterns

I also only applied it on the [#factory](https://beta.infinum.com/handbook/books/rails/development-practices/design-patterns#factory) chapters, cause it'll take some time to restructure all patterns in this way (everything inside `<details>` should be HTML, not MD anymore).
Let me know if this is the way we should go and I'll apply it to the rest.

melcha commented:
@cilim this looks much more readable! Could you somehow highlight the summary content, so we know it's clickable?

cilim commented:
@melcha something discreet

<img width="830" alt="Screenshot 2020-06-16 at 15 46 22" src="https://user-images.githubusercontent.com/4703945/84782759-b3adb280-afe8-11ea-9b18-0e15c7cac65b.png">


or something more fun? 😄 

<img width="656" alt="Screenshot 2020-06-16 at 15 45 42" src="https://user-images.githubusercontent.com/4703945/84782706-a55f9680-afe8-11ea-945e-be7ca266d8e4.png">

cukoreva commented on Development practices/Design patterns.md:
```diff
@@ -1,12 +1,274 @@
+Simplify record operations by keeping logic in them.
+
+#### Problem
+* Your controllers are huge and you want to make them more readable by extracting some of its logic, but don't want to clutter your models.
```
```suggestion
* Your controllers are huge and you want to make them more readable by extracting some of their logic, but you don't want to clutter your models.
```

cukoreva commented on Development practices/Design patterns.md:
```diff
@@ -1,12 +1,274 @@
+#### Problem
+* Your controllers are huge and you want to make them more readable by extracting some of its logic, but don't want to clutter your models.
+* Your opinion is that models shouldn't have validations.
+* You have a virtual attribute that works only on a couple of places, but doesn't make sense to add `attr_accessor` to the model.
```
```suggestion
* You have a virtual attribute that works only in a couple of places, but it doesn't make sense to add `attr_accessor` to the model.
```

cukoreva commented on Development practices/Design patterns.md:
```diff
@@ -1,12 +1,274 @@
+Instead of returning null, or some odd value, return a Special Case that has the same interface as what the caller expects.
+
+#### Problem
+You have a lot of conditionals scattered around many classes or view files that ask the same thing. It's usually a low-level error handling part that covers calls to `nil`s.
```
```suggestion
You have a lot of conditionals scattered around many classes or view files that check the same thing. It's usually a low-level error handling part that covers calls to `nil`s.
```

cukoreva commented on Development practices/Design patterns.md:
```diff
@@ -1,12 +1,274 @@
+Concentrate the core logic of a request operation into a separate object.
+
+#### Problem
+You have a complex set of operations that need to be executed in order synchronously or asynchronously.
```
```suggestion
You have a complex set of operations that need to be executed in a sequence synchronously or asynchronously.
```

cukoreva commented on Development practices/Design patterns.md:
```diff
@@ -1,12 +1,274 @@
+_Note_: Service objects can easily become a code smell if not handled properly. If you find yourself with a service object that has too many methods and/or becomes generally unreadable and hard to understand, you might have to find a [bounded context](https://blog.carbonfive.com/bring-clarity-to-your-monolith-with-bounded-contexts/) or use one of the design patterns mentioned in this chapter.
+
+### Builder
+Unlike the factories, this design pattern doesn't have a common interface. That makes it possible to different products, using the same construction process.
```
```suggestion
Unlike the factories, this design pattern doesn't have a common interface. That makes it possible to create different products, using the same construction process.
```

cukoreva commented on Development practices/Design patterns.md:
```diff
@@ -1,12 +1,274 @@
+Read more about *Prototype* [here](https://refactoring.guru/design-patterns/prototype).
+
+[Ruby example](https://refactoring.guru/design-patterns/prototype/ruby/example).
+_Note_: The Ruby example uses the usual Marshalling hack to make a deep copy. But it's rather slow and inefficient, therefore, in real applications, use a special gem (eg [deep_cloneable](https://github.com/moiristo/deep_cloneable))
```
```suggestion
_Note_: The Ruby example uses the usual Marshalling hack to make a deep copy. But it's rather slow and inefficient, therefore, in real applications, use a special gem (e.g. [deep_cloneable](https://github.com/moiristo/deep_cloneable))
```

cukoreva commented on Development practices/Design patterns.md:
```diff
@@ -1,12 +1,274 @@
+_Note_: The Ruby example uses the usual Marshalling hack to make a deep copy. But it's rather slow and inefficient, therefore, in real applications, use a special gem (eg [deep_cloneable](https://github.com/moiristo/deep_cloneable))
+
+### Singleton
+Ensure a class to only have 1 object.
```
```suggestion
Ensure a class has only one object.
```

cukoreva commented on Development practices/Design patterns.md:
```diff
@@ -1,12 +1,274 @@
+[Ruby example](https://refactoring.guru/design-patterns/adapter/ruby/example).
+
+### Bridge
+Split a large class or a set of closely related classes into two separate hierarchies
```
```suggestion
Split a large class or a set of closely related classes into two separate hierarchies.
```

cukoreva replied:
Please check if there are other places where the full stop is missing.

cukoreva commented on Development practices/Design patterns.md:
```diff
@@ -1,12 +1,274 @@
+Split a large class or a set of closely related classes into two separate hierarchies
+
+#### Problem
+You're planning to add new subclasses which would complicate the class hierarchy.
```
```suggestion
You're planning to add new subclasses, which would complicate the class hierarchy.
```

cukoreva commented on Development practices/Design patterns.md:
```diff
@@ -1,12 +1,274 @@
+
+#### Problem
+You need a method that is using an ActiveRecord object attribute during the operation.
+You don't want to add the new method to the model cause it's only used in 1 place.
```
```suggestion
You don't want to add the new method to the model cause it's only used in one place.
```
Number 0-10 are normally written with words (when part of a text).

cukoreva commented on Development practices/Design patterns.md:
```diff
@@ -1,12 +1,274 @@
+_Note_: The [draper](https://github.com/drapergem/draper) gem does exactly what you want.
+
+### Facade
+Provides a well documented, simplified and limited interface to a more complex system.
```
```suggestion
Provides a well-documented, simplified and limited interface to a more complex system.
```

cukoreva commented on Development practices/Design patterns.md:
```diff
@@ -1,12 +1,274 @@
+Define a skeleton of an algorithm in a base class and let subclasses override the steps without changing the overall algorithm’s structure.
+
+#### Problem
+You have many classes that solve the same semantical issue (importing and data processing from files, JSON/CSV/Excel) and have the same algorithm, but the parsing is different.
```
```suggestion
You have many classes that solve the same semantic issue (importing and data processing from files, JSON/CSV/Excel) and have the same algorithm, but the parsing is different.
```

cukoreva commented on Development practices/Design patterns.md:
```diff
@@ -1,12 +1,274 @@
+[Ruby example](https://refactoring.guru/design-patterns/strategy/ruby/example).
+
+### Template Method
+Define a skeleton of an algorithm in a base class and let subclasses override the steps without changing the overall algorithm’s structure.
```
```suggestion
Define a skeleton of an algorithm in a base class and let subclasses override the steps without changing the overall algorithm structure.
```

cukoreva commented on Development practices/Design patterns.md:
```diff
@@ -1,12 +1,274 @@
+Turn a set of behaviours into objects and make them interchangeable inside the original context object.
+
+#### Problem
+Your application requirements changed and now you have to communicate with a 2nd API that returns same type of data as the 1st one.
```
```suggestion
Your application requirements changed and now you have to communicate with a second API that returns the same type of data as the first one.
```

melcha commented on Development practices/Design patterns.md:
```diff
@@ -1,12 +1,274 @@
+  <summary>Solution</summary>
+  Read more about <b>Abstract factory</b> <a href="https://refactoring.guru/design-patterns/abstract-factory">here</a>.
+  <br>
+  <a href="https://refactoring.guru/design-patterns/factory-method/ruby/example">Ruby example</a>.
```
```suggestion
  <a href="https://refactoring.guru/design-patterns/abstract-factory/ruby/example">Ruby example</a>.
```

melcha commented:
@cilim this looks much more readable! Could you somehow highlight the summary content, so we know it's clickable?

> @melcha something discreet
> 
> <img alt="Screenshot 2020-06-16 at 15 46 22" width="830" src="https://user-images.githubusercontent.com/4703945/84782759-b3adb280-afe8-11ea-9b18-0e15c7cac65b.png">
> 
> or something more fun? 😄
> 
> <img alt="Screenshot 2020-06-16 at 15 45 42" width="656" src="https://user-images.githubusercontent.com/4703945/84782706-a55f9680-afe8-11ea-945e-be7ca266d8e4.png">

Let's go with the first one, but without the underline

vr4b4c commented:
This is chapter is too broad with too many patterns covered. I would simplify it by dividing it into two sections. 

*Generic/standard/classic design patterns*
• general purpose of design patterns (why)
• giant link to code guru for more details about each pattern (how)

*Rails specific patterns*
• Form object (in conjunction with ActiveType)
• Query object
• Service object
• Decorator (Draper, SimpleDelegator, Forwardable)

cilim commented:
@vr4b4c since the majority of people that reviewed the PR agree with style > readability (in context of hiding the `Problem` and `Solution` detail sections which I intended to simplify finding what you need) I believe that this is an ok argument to keep all general design patterns just behind a huge link to refactoring.guru, and keep the Rails specific patterns as they are.

What do you think @melcha @d4be4st ?

melcha commented:
Actually I do agree with @vr4b4c. Initially I thought we would cover only Rails related patterns we use. 
There's no need to bloat the chapter with references to selected general purpose design patterns (we don't link all of them) if we can just link to refactoring guru, they've covered everything and the site looks great.

cilim commented:
Changes applied with a fixup.

Deployed to staging. Check it out: https://beta.infinum.com/handbook/books/rails/development-practices/design-patterns

vr4b4c commented:
Structurally, exactly on point with what I was suggesting. Stylewise, I don't think there is a need for collapsible Problem/Solution sections.

cilim commented:
>Structurally, exactly on point with what I was suggesting. Stylewise, I don't think there is a need for collapsible Problem/Solution sections.

@d4be4st @melcha you were rooting for those, what do you say? Maybe it won't look bad now that there's much less content.

melcha commented:
Only because there was a lot of content and no support for h4. We can remove it now.

cilim commented:
All done.

cukoreva commented on Development practices/Design patterns.md:
```diff
@@ -1,12 +1,74 @@
+* You have a complex SQL query that is used in multiple callers
+
+#### Solution
+Read more [here](https://medium.flatstack.com/query-object-in-ruby-on-rails-56ea434365f0)..
```
```suggestion
Read more [here](https://medium.flatstack.com/query-object-in-ruby-on-rails-56ea434365f0).
```

cilim replied:
Good 👁️ !

d4be4st commented on Development practices/Design patterns.md:
```diff
@@ -1,12 +1,74 @@
+* logic that is tightly coupled with primitives (attributes with behaviour - methods)
+
+#### Solution
+Read more [here](https://revs.runtime-revolution.com/value-objects-in-ruby-on-rails-9df64bc8db34).
```
would mention https://dry-rb.org/gems/dry-initializer/3.0/ here

d4be4st commented on Development practices/Design patterns.md:
```diff
@@ -1,12 +1,74 @@
+## Rails patterns
+
+### Factory
+Class that has one creation method with a large conditional.
```
A class that creates other classes? :D

cilim replied:
Just saw this!

I like it somewhat more descriptive :D It technically doesn't only need to return a class, but can also return an object.

d4be4st commented on Development practices/Design patterns.md:
```diff
@@ -1,12 +1,74 @@
+Read more [here](https://www.sihui.io/design-pattern-factory/).
+
+### Form object
+Simplify record operations by keeping logic in them.
```
I would refraze this somehow...
The old definition was somehow better for me.

cilim replied:
@d4be4st last changes: https://github.com/infinum/rails-handbook/pull/117/commits/b2ed21578f5a003fc562f5f30fefc32b17318ee6#diff-f5cea16a5f762555e1d1efbf4c751f96R24

Let me know if you agree with @melcha and we can merge it.

cilim commented:
@infinum/backend-team any other volunteers eager to comment? :D Or planning to comment...

It's somewhat unproductive to revisit this branch after every N days when someone comments something new 🙉

melcha commented:
You can merge it 👍
